File: 101_symmetric_tree/recursive.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        """
        Complexity: O(n)
        Space: O(n)
        """
        if root == None:
            return True
        def sub_isSymmetric(right, left):
            if right == None and left == None:
                return True
            if right == None or left == None:
                return False
            if right.val == left.val and sub_isSymmetric(right.right, left.left) and sub_isSymmetric(right.left, left.right):
                return True
            else:
                return False
        return sub_isSymmetric(root.right, root.left)

File: 101_symmetric_tree/recursive.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool is_subSymmetric(TreeNode* right, TreeNode* left) {
        if(right == NULL && left == NULL) return true;
        else if(right == NULL || left == NULL) return false;
        else if(right->val == left->val && is_subSymmetric(right->right, left->left) && is_subSymmetric(right->left, left->right)) return true;
        else return false;
    }
    bool isSymmetric(TreeNode* root) {
        if(root == NULL) return true;
        return is_subSymmetric(root->right, root->left);
    }
};
